docs/RationaleHWArith.md

```diff
@@ -0,0 +1,71 @@
+    - [Q&A](#qa)
+  - [Bit Width Rules](#bit-width-rules)
+  - [Lowering](#lowering)
+
+## Introduction
+
+The `hwarith` dialect provides a collection of operations that define typical integer arithmetic operations which are bitwidth aware.
+This bitwidth awareness is expressed through return type inference rules that, based on the types of an operation and its input operands, infers the type of the result operand.
+
+In general, the bit width inference rules are designed to model SystemVerilog semantics. However, unlike SystemVerilog, the aim of this dialect is to provide a strongly-typed hardware arithmetic library with the width inference rules for every operator explained in a clear and understandable way to avoid ambiguity.
+
+To serve this goal of being strongly typed, the operations of the dialect explicitly does not accept signless types. As such, only `ui#`, `si#` types are allowed. A casting operation is provided to serve signedness casting as well as truncation and padding.
+
+Below we try to capture some common questions on the capabilities of such a dialect, and how we've addressed/think about the issue.
+
+### Q&A
+* **Q:** Does this support n-ary operations?
+    * n-ary operations in this dialect would have to solve the issue of how width rules in expressions such as the following are handled:
+    `%0 = hwarith.add %0, %1, %2 : si6, si5, si4
+    In the short term, we invision this dialect to be considering strictly binary operations. Defining width inference rules for these is simpler and fits the immediate usecase. If, in the future, a need for n-ary operations comes up and is motivated clearly (e.g. for optimization purposes), the dialect should be adapted to support it. In this case, we expect n-ary operation rules to be a superset of those defined for binary operations.
+* **Q:** Does this support 0-width operands?
+    * 0-width values might arise from arithmetic rules which reduce the bit width of an expression wrt. the operands to that expression. One case where such rule _may_ apply is in the implementation of modulo operations.  
+    We refrain from adding support at this point in time, since support for 0-width values in the remainder of the RTL dialects is, at time of writing, lacking/undefined. Once support is added in these downstream dialect, 0-width support in `hwarith` should be reconsidered.
+* **Q:** Does this support width inferred types?
+    * Relying on width inference is relevant when referencing results of other width-inferred value. Without this, a user/frontend must itself know and apply width inference rules before generating the IR. Having width-inferred types will be convenient not only for generating IR, but also to leave room for width inference rules and optimizations to apply recursively. As an example:
+    ```mlir
+    %1 = hwarith.add %a, %b : ui4, ui4
+    %2 = hwarith.add %c, %1 : ui3, ui // %1 is an unsigned integer of inferred width
+    ```
+    @TODO: (When) do we want this?
```

Comment:
Interesting, but this is only an issue for humans/frontends emitting textual IR, as otherwise the ops' builders would handle the type inference, correct?

Reply:
I don't think this is wise to do -- MLIR expects to be able to resolve types on a per-op basis independent of other ops. Rather unfortunate that we can't do this.

There may be some merit in creating an "inferred width int" type at some point. I think FIRRTL has something similar, and IIRC I've heard complaints about having to do type inference. So that's not a problem I think we should be addressing in this dialect.

Reply:
An uninfered type has to be a real type.